server/src/main/java/org/apache/druid/catalog/model/table/ClusterKeySpec.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.catalog.model.table;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.Nullable;
+
+import java.util.Objects;
+
+public class ClusterKeySpec
+{
+  private final String expr;
+  private final boolean desc;
```

Comment:
FWIU, I think clustering is a concept which is currently sorting but might be extended in future to hashing or space filling curves, trees as well. So, I thought that instead of hardcoding an attribute in the spec we could make it extendible. wdyt? 

For column based indexes, I completely agree with you that those can be per-column properties

Reply:
I see the point. Do we have a design for the new kinds of clustering at the column level? For example, instead of sorting a key, might we hash it? Or, would we say, "do sort it, but use XYZ sort instead of ABC sort?" How would we represent this in SQL? At present, `CLUSTER BY` looks like a `SORT` spec, I believe.

If we have a design, then I can update this item to match.

If, on the other hand, these other items have to do with _indexing_ of columns (not the segment sort order), then column properties would be the way to go.

Reply:
Is there any harm is making this extensible even if can be mapped to desc/asc boolean for foreseable future? Would it require a lot of code churn to update this? I agree with @rohangarg that would be good to make this flexible.

Reply:
Exactly what would the extension be? Clustering is sorting. Sorting is either ascending or descending. Let's say we had a hash index. Could we have a "hash clustering"? I'm not sure what that means.

If we want to extend, then let's wait for the new innovation. Then, if we, say, provide a hash table as a an alternative to sorting, we provide a `hashing` property that defines the attributes of that alternative, and say that you can provide either clustering or ordering.

If there is something in the works now, then pease point me to the spec and I'll be happy to model it in the catalog.

Reply:
One other thought. Note that the asc/desc option is designed to default to the most frequent choice:

```java
  public ClusterKeySpec(
      @JsonProperty("column") String expr,
      @JsonProperty("desc") @Nullable Boolean desc
  )
  {
    this.expr = expr;
    this.desc = desc != null && desc;
  }
```

Let's say we wanted to be SQL compliant and let the user specify `NULLS HIGH` or `NULLS LOW`. We would use the same trick which would be nicely backward compatible:

```java
  public ClusterKeySpec(
      @JsonProperty("column") String expr,
      @JsonProperty("desc") @Nullable Boolean desc,
      @JsonProperty("nullsHigh") @Nullable Boolean nullsHigh
  )
  {
    this.expr = expr;
    this.desc = desc != null && desc;
    this.nullsHigh = nullsHigh != null & nullsHigh;
  }
```